### PR TITLE
feat(pipettes): add GPIO EXTI interrupt support for the tip sense GPIOs

### PR DESCRIPTION
**Pull request by Laura-Danielle**
## Overview

We want to decide that a tip has been detected based on an EXTI interrupt. This PR adds the piping in HAL to support an EXTI interrupt. While making these changes, I also edited the code on the pressure sensor to support two data ready lines for the 96 channel. We'll also want to do this for DVT 8 channel pipettes since there are now two separate data ready lines.

## Test Plan

- [x] Test that the pressure sensor still functions normally on the single/eight channel

### CODE REVIEW

File: pipettes/firmware/utility_gpio.c [hunk 1]
```diff
@@ -166,9 +169,9 @@ static void data_ready_gpio_init() {
         GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
         HAL_GPIO_Init(hardware.port, &GPIO_InitStruct);
 
-        /* EXTI interrupt init*/
-        HAL_NVIC_SetPriority(exti_line, 10, 0);
-        HAL_NVIC_EnableIRQ(exti_line);
+        // /* EXTI interrupt init*/
```
Comment (fsinapi) [inline pipettes/firmware/utility_gpio.c:172]: Guessing this should be uncommented before merging?

File: pipettes/firmware/stm32g4xx_it.c [hunk 2]
```diff
@@ -118,12 +118,37 @@ void DebugMon_Handler(void) {}
 /*  file (startup_stm32g4xxxx.s).                                             */
 /******************************************************************************/
 
+void EXTI2_IRQHandler(void) {
+    EXTI->PR1 = EXTI_LINE_2;
```
Comment (fsinapi) [inline pipettes/firmware/stm32g4xx_it.c:122]: Similar to checking `PR1`, can we clear using `__HAL_GPIO_EXTI_CLEAR_FLAG`?

File: pipettes/firmware/hardware_config.c [hunk 3]
```diff
@@ -33,23 +33,45 @@ static PipetteHardwarePin get_gpio_ht(PipetteHardwareDevice device) {
...
             // PC6 -> GPIO_EXTI6 (EXTI9_5_IRQn)
             // PC11 -> GPIO_EXTI11 (EXTI15_10_IRQn)
-            return EXTI15_10_IRQn;
+            return EXTI15_10_IRQn | EXTI9_5_IRQn;
+        case pipette_gpio_tip_sense:
+        default:
+            // tip sense
+            // PC7 -> GPIO_EXTI6 (EXTI9_5_IRQn)
+            // PC12 -> GPIO_EXTI11 (EXTI15_10_IRQn)
+            return EXTI15_10_IRQn | EXTI9_5_IRQn;
```
Comment (fsinapi) [inline pipettes/firmware/hardware_config.c:61]: I don't think that this is going to give you the behavior you want. The `IRQn` enums are not single-bit values, but instead indices into the NVIC bitfields. Returning two IRQn values OR'd together is probably going to end up enabling a completely different IRQ.

**Review by fsinapi — CHANGES_REQUESTED** [review]
Had a few comments but the big one is how `get_interrupt_line_ht` tries to return two different IRQ lines, I'm pretty sure we're going to have to use a different approach there.

**Comment by Laura-Danielle** [comment]
Yeah I needed to test this on the firmware to see if that would actually work. I know you can combine GPIOs but it wasn't clear to me in the NVIC documentation whether the same thing would work or not. I can just change the implementation (and probably make the code more readable by doing so.)

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

**Review by caila-marashaj — APPROVED** [review]
Exciting!!

**Comment by Laura-Danielle** [comment]
Tested liquid sensing, auto calibration and the new tip sense functionality on a robot. Everything is functioning as expected ✔️

**Pull request merged**